### PR TITLE
Added tryAccept function (accept without throwing)

### PR DESCRIPTION
**Pull request by WebFreak001**
accept throws a SocketAcceptException when it is non-blocking and there is no connection which is bad for performance. tryAccept simply returns null instead of throwing

### CODE REVIEW

File: std/socket.d [hunk 1]
```diff
@@ -2808,13 +2808,26 @@ public:
     /**
      * Accept an incoming connection. If the socket is blocking, $(D accept)
      * waits for a connection request. Throws $(D SocketAcceptException) if
-     * unable to _accept. See $(D accepting) for use with derived classes.
+     * unable to accept. See $(D accepting) for use with derived classes.
      */
```
Comment (wilzbach) [inline std/socket.d]: Needs "Returns"

File: std/socket.d [hunk 2]
```diff
@@ -2808,13 +2808,26 @@ public:
...
+    {
+        auto newSocket = tryAccept();
+        if (newSocket is null)
+            throw new SocketAcceptException("Unable to accept socket connection");
+        return newSocket;
+    }
+
+    /**
+     * Accept an incoming connection. If the socket is blocking, $(D tryAccept)
+     * waits for a connection request. Returns $(D null) if unable to accept.
```
Comment (wilzbach) [inline std/socket.d]: Please use "Returns:" (i.e. e Ddoc format) and state that a `Socket` object is returned by default (I know it's obvious, but we're trying to make the docs better and more intuitive).

File: std/socket.d [hunk 3]
```diff
@@ -2808,13 +2808,26 @@ public:
...
+     * See $(D accepting) for use with derived classes.
```
Comment (wilzbach) [inline std/socket.d]: I know that this is part of the existing docs, but it's there any reason for `accepting` not being a link?

File: std/socket.d [hunk 4]
```diff
@@ -2808,13 +2808,26 @@ public:
     /**
      * Accept an incoming connection. If the socket is blocking, $(D accept)
      * waits for a connection request. Throws $(D SocketAcceptException) if
-     * unable to _accept. See $(D accepting) for use with derived classes.
+     * unable to accept. See $(D accepting) for use with derived classes.
```
Comment (CyberShadow) [inline std/socket.d]: Why this change?

**Comment by CyberShadow** [comment]
> accept throws a SocketAcceptException when it is non-blocking and there is no connection which is bad for performance.

Generally the non-blocking flow is to add a listening socket to a SocketSet and `select` loop, then only call `accept` once `select` indicates it's readable.

Do you have a use case where the above is not applicable?

File: std/socket.d (hunk 4 above)
Comment (WebFreak001) [inline std/socket.d]: oops thought it was a typo, but I see why it was there now

**Comment by WebFreak001** [comment]
Uh I rarely ever used `select` and `SocketSet` but the advantage with this function is that it looks much more simple and more intuitive, plus it integrates much smoother with Fibers.

**Comment by CyberShadow** [comment]
Can you give me an example of how you would use this function in your application?

**Comment by WebFreak001** [comment]
```d
// simplified but a bit like this
void acceptFiber()
{
  while (true)
  {
    auto client = server.tryAccept;
    if (client !is null)
      spawnFiber(new ClientFiber(client));
    Fiber.yield();
  }
}
```

I know it is possible with SocketSet and `select` too but I simply find this more elegant

**Comment by CyberShadow** [comment]
> // simplified but a bit like this

That code is not great because it will spin consuming 100% CPU all the time.

> Uh I rarely ever used select and SocketSet

If you are using fibers and non-blocking sockets, you pretty much always also want to have an event loop. Generally, the event loop holds the responsibility of notifying the fibers when a connection socket is readable, or a listening socket is ready to accept a new connection.

I suggest that you have a look at how various networking libraries implement this, and try it out in practice in your own application, before submitting a patch to Phobos.

OK to close?

**Comment by WebFreak001** [comment]
hm ok sure you can close this, I am currently writing an implementation with SocketSet to try it out because the CPU was so busy, I simply didn't do so before because it felt a bit hacky always readding the sockets (tip for other readers, there is a nice example doing a simple server with SocketSet [here](https://github.com/dlang/dmd/blob/master/samples/listener.d)).

Also I'm just gonna push the documentation updates for the sake of completion though
